cmd/ndm_daemonset/probe/mountprobe.go

```diff
@@ -106,3 +148,85 @@ func (mp *mountProbe) FillBlockDeviceDetails(blockDevice *blockdevice.BlockDevic
 		blockDevice.FSInfo.FileSystem = basicMountInfo.FileSystem
 	}
 }
+
+func (mp *mountProbe) setupEpoll() error {
+	ep, err := epoll.New()
+	if err != nil {
+		return err
+	}
+	mp.epoll = &ep
+	return ep.AddWatcher(epoll.Watcher{
+		FileName:   mp.mountsFileName,
+		EventTypes: []epoll.EventType{epoll.EPOLLERR, epoll.EPOLLPRI},
+	})
+}
+
+func (mp *mountProbe) listen() {
+	eventCh, err := mp.epoll.Start()
+	if err != nil {
+		klog.Errorf("error while starting epoll: %v", err)
+		return
+	}
+	defer mp.epoll.Stop()
+	klog.Info("started mount change detection loop")
+	defaultMsg := controller.EventMessage{
+		Action:          string(MountEA),
+		Devices:         nil,
+		AllBlockDevices: true,
+	}
+
+	for range eventCh {
+		// regenerate mounts table and get the changes
+		newMountTable, err := mp.newMountTable()
+		if err != nil {
+			klog.Error("failed to generate mounts table.")
+			mp.destination <- defaultMsg
+		}
+		mtDiff := libmount.GenerateDiff(mp.mountTable, newMountTable)
+		mp.mountTable = newMountTable
+		mp.processDiff(mtDiff)
+	}
+}
+
+func (mp *mountProbe) newMountTable() (*libmount.MountTab, error) {
+	return libmount.NewMountTab(libmount.FromFile(mp.mountsFileName,
+		libmount.MNT_FMT_FSTAB),
+		libmount.WithAllowFilter(libmount.SourceContainsFilter("/dev/")),
+		libmount.WithDenyFilter(libmount.SourceFilter("overlay")),
+		libmount.WithDenyFilter(libmount.TargetContainsFilter("/var/lib/kubelet/pod")))
```

Comment:
Do we need to add `/var/lib/docker` also here, or is it under the overlay target.

Reply:
Also this item also needs to be filtered out, I am seeing the below messages in one of the nodes in my cluster.
```
mount      /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.UsP2mq /dev/sda5[/usr/bin/runc] ext4   rw,relatime,errors=remount-ro
remount    /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.UsP2mq /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
umount     /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.UsP2mq /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
mount      /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.gXCAMA /dev/sda5[/usr/bin/runc] ext4   rw,relatime,errors=remount-ro
remount    /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.gXCAMA /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
umount     /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.gXCAMA /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
mount      /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.j7g4kS /dev/sda5[/usr/bin/runc] ext4   rw,relatime,errors=remount-ro
remount    /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.j7g4kS /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
umount     /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.j7g4kS /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
mount      /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.W6lDe6 /dev/sda5[/usr/bin/runc] ext4   rw,relatime,errors=remount-ro
remount    /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.W6lDe6 /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
umount     /run/docker/runtime-runc/moby/2fa14e717b0d77a91d990b02158fb78f9a7c64dc61d1064507c3ff26c4812779/runc.W6lDe6 /dev/sda5[/usr/bin/runc] ext4   ro,relatime,errors=remount-ro
```

The corresponding processing takes place in the mount probe, but gets ignored by os disk filter since its happening on the root disk (`/dev/sda5` is the root disk). We need to move this filtering into the mountab filter.

Reply:
so there should be a filter for target `/run/docker/` too?

Reply:
Yes.